crates/bevy_render/src/view/visibility/mod.rs

```diff
@@ -279,74 +231,97 @@ impl Plugin for VisibilityPlugin {
     }
 }
 
-/// Calculates [`Aabb`]s for [`Entities`](Entity) with [`Mesh`]es.
-pub fn calculate_bounds(
-    mut commands: Commands,
-    meshes: Res<Assets<Mesh>>,
-    without_aabb: Query<(Entity, &Handle<Mesh>), (Without<Aabb>, Without<NoFrustumCulling>)>,
-    mut entity_mesh_map: ResMut<EntityMeshMap>,
-) {
-    for (entity, mesh_handle) in &without_aabb {
-        if let Some(mesh) = meshes.get(mesh_handle) {
-            if let Some(aabb) = mesh.compute_aabb() {
-                entity_mesh_map.register(entity, mesh_handle);
-                commands.entity(entity).insert(aabb);
-            }
-        }
-    }
-}
-
 /// Updates [`Aabb`]s for [`Entities`](Entity) with [`Mesh`]es. This includes `Entities` that have
 /// been assigned new `Mesh`es as well as `Entities` whose `Mesh` has been directly mutated.
 ///
 /// To opt out of bound calculation for an `Entity`, give it the [`NoAabbUpdate`] component.
-///
-/// NOTE: This system needs to remove entities from their collection in
-/// [`EntityMeshMap`] whenever a mesh handle is reassigned or an entity's mesh handle is
-/// removed. This may impact performance if meshes with many entities are frequently
-/// reassigned/removed.
 pub fn update_bounds(
     mut commands: Commands,
     meshes: Res<Assets<Mesh>>,
-    mut mesh_reassigned: Query<
-        (Entity, &Handle<Mesh>, &mut Aabb),
-        (
-            Changed<Handle<Mesh>>,
-            Without<NoFrustumCulling>,
-            Without<NoAabbUpdate>,
-        ),
-    >,
-    mut entity_mesh_map: ResMut<EntityMeshMap>,
+    mut queries: ParamSet<(
+        Query<(&Handle<Mesh>, &Aabb), Changed<Aabb>>,
+        Query<
+            (
+                Entity,
+                &Handle<Mesh>,
+                Changed<Handle<Mesh>>,
+                Option<&mut Aabb>,
+            ),
+            (Without<NoFrustumCulling>, Without<NoAabbUpdate>),
+        >,
+    )>,
+    mut mesh_aabb_map: ResMut<MeshAabbMap>,
     mut mesh_events: EventReader<AssetEvent<Mesh>>,
-    entities_lost_mesh: RemovedComponents<Handle<Mesh>>,
+    mut changed_aabbs: Local<HashSet<Handle<Mesh>>>,
+    mut inserted_aabb_components: Local<HashSet<Handle<Mesh>>>,
 ) {
-    for entity in entities_lost_mesh.iter() {
-        entity_mesh_map.deregister(entity);
+    // Identify Aabbs added to / updated on entities with Handle<Mesh> outside of this system. The
+    // goal here is to avoid recomputing Aabbs for Meshes that have already been computed
+    // elsewhere.
+    for (mesh_handle, aabb) in queries.p0().iter() {
+        // If the Aabb component was inserted using Commands on the previous run of this system,
+        // do not consider it as Changed for this run.
```

Comment:
Couldn't this miss updates on consecutive frames?

Reply:
I don't understand the question. Could you elaborate a bit on what you think could be a problem case that this misses?

Reply:
Note that I clear `inserted_aabb_components` after executing this loop.

Reply:
What if, on frame N, the `Mesh` is updated, and then on frame N+1, the `Aabb` is updated by another system? It seems the update to the `Aabb` would be ignored, and so there could be two entities with the same `Handle<Mesh>` and yet different `Aabb`s.

It's not clear what this system is trying to accomplish - it seems either it should be entirely responsible for updating the `Aabb` when the `NoAabbUpdate` component is missing, and hence no need to check if the `Aabb` has changed, or it should not update the `Aabb` at all when `NoAabbUpdate` is present.

Reply:
The problem is that NoAabbUpdate is a component, and I have made an assumption that an Aabb belongs to a Mesh asset but is also a component. I wanted to at least try to respect and indeed avoid calculating Aabbs for meshes if they have been manually added. For example, gltf stores min/max values of things including vertex positions which allows for very fast Aabb creation. I don’t want to ignore that and overwrite that Aabb for a glTF Mesh when it is already correct. Also, later I expect we will store Aabbs in metadata and will want to load them and not overwrite them. This should be the common case - either it has an Aabb at spawn time, or it needs one calculating, and it isn’t modified. But we also need to support runtime-modified Meshes and then we have to recompute unless the modification process has a faster way of updating the Aabb.

I recognise the corner case you point out, and maybe there is a way to address it by moving some of this logic to another system. I’ll have to think through it carefully.